File: windows/CursorProvider.cs
using System.IO;
using System.Net.Http;
using System.Text.Json;
using static Brink.L10n;

namespace Brink;

/// Reads the Cursor CLI's session JWT and queries Cursor's `usage-summary`
/// endpoint — the source behind the "You've used N% of your included usage"
/// message in the Cursor app.
///
/// Windows credential lookup (the macOS build reads the Keychain item
/// "cursor-access-token" / "cursor-user"):
///  1) %USERPROFILE%\.cursor\cli-config.json — authInfo.accessToken, when the
///     CLI stores the token in the config itself
///  2) Credential Manager generic entries "cursor-access-token" / "cursor-user"
/// The numeric userId always comes from cli-config.json (authInfo.userId) and
/// prefixes the session cookie as "<userId>::<jwt>".
///
/// Brink never touches the refresh token — refreshing would rotate it and log
/// the user out of `cursor-agent`. Cursor exposes no public usage API, so every
/// field is probed defensively.
public class CursorProvider : IUsageProvider
{
    public string Id => "cursor";

    private const string UsageUrl = "https://cursor.com/api/usage-summary";

    private static string? _tokenCache;
    private static string? _userIdCache;

    public async Task<ProviderSnapshot> FetchAsync()
    {
        var snap = new ProviderSnapshot { Id = Id, Name = "Cursor" };
        var token = LoadToken();
        var userId = LoadUserId();
        if (token == null || userId == null)
            return ClaudeProvider.DemoSnapshot("Cursor",
                L("Cursor CLI credentials not found — run `cursor-agent login`"));

        try
        {
            var (data, status) = await RequestUsageAsync(token, userId);
            if (status is 401 or 403)
            {
                // cursor-agent rotated its token: drop cache, re-read once.
                ClearCache();
                var fresh = LoadToken();
                if (fresh != null && fresh != token)
                    (data, status) = await RequestUsageAsync(fresh, userId);
            }
            if (status is 401 or 403)
            {
                snap.Error = L("Unauthorized — run `cursor-agent login` to refresh");
                return snap;
            }
            if (status != 200)
            {
                snap.Error = L("HTTP %d", status);
                return snap;
            }
            snap.Windows = ParseUsage(data);
            snap.UpdatedAt = DateTime.Now;
            if (snap.Windows.Count == 0) snap.Error = L("No usage data in response");
            return snap;
        }
        catch (Exception e)
        {
            snap.Error = e.Message;
            return snap;
        }
    }

    private static async Task<(string Data, int Status)> RequestUsageAsync(string token, string userId)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, UsageUrl);
        // Session-cookie auth: "WorkosCursorSessionToken=<userId>::<jwt>", ":" percent-encoded.
        request.Headers.TryAddWithoutValidation("Cookie",
            $"WorkosCursorSessionToken={userId}%3A%3A{token}");
        request.Headers.TryAddWithoutValidation("User-Agent", "Brink/1.0");
        using var response = await ClaudeProvider.Http.SendAsync(request);
        var body = await response.Content.ReadAsStringAsync();
        return (body, (int)response.StatusCode);
    }

    // MARK: Parsing

    /// usage-summary shape (fields probed defensively):
    /// { individualUsage: { plan: { totalPercentUsed, autoPercentUsed, apiPercentUsed, ... },
    ///                      onDemand: { enabled, limit, used } },
    ///   isUnlimited, billingCycleEnd }
    /// Labels are stored as keys and localized at render time.
    public static List<UsageWindow> ParseUsage(string data)
    {
        var windows = new List<UsageWindow>();
        try
        {
            using var doc = JsonDocument.Parse(data);
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return windows;
            var reset = ClaudeProvider.GetIsoDate(root, "billingCycleEnd");

            // Unlimited plans have no meaningful percentage to show.
            if (GetFlag(root, "isUnlimited"))
            {
                windows.Add(new UsageWindow { Label = "Included usage", UsedPercent = 0, ResetsAt = reset });
                return windows;
            }

            JsonElement plan = default, onDemand = default;
            bool hasPlan = false, hasOnDemand = false;
            if (root.TryGetProperty("individualUsage", out var individual)
                && individual.ValueKind == JsonValueKind.Object)
            {
                hasPlan = individual.TryGetProperty("plan", out plan)
                    && plan.ValueKind == JsonValueKind.Object;
                hasOnDemand = individual.TryGetProperty("onDemand", out onDemand)
                    && onDemand.ValueKind == JsonValueKind.Object;
            }

            if (hasPlan)
            {
                // Headline window — total included usage (matches the app's banner).
                if (ClaudeProvider.GetNumber(plan, "totalPercentUsed") is double total)
                    windows.Add(new UsageWindow { Label = "Included usage", UsedPercent = total, ResetsAt = reset });
                if (ClaudeProvider.GetNumber(plan, "autoPercentUsed") is double auto)
                    windows.Add(new UsageWindow { Label = "Auto", UsedPercent = auto, ResetsAt = reset });
                if (ClaudeProvider.GetNumber(plan, "apiPercentUsed") is double api)
                    windows.Add(new UsageWindow { Label = "API", UsedPercent = api, ResetsAt = reset });
            }

            // On-demand / usage-based spend, only when the user has enabled it.
            if (hasOnDemand && GetFlag(onDemand, "enabled")
                && ClaudeProvider.GetNumber(onDemand, "used") is double used
                && ClaudeProvider.GetNumber(onDemand, "limit") is double limit && limit > 0)
            {
                windows.Add(new UsageWindow
                {
                    Label = "On-demand",
                    UsedPercent = used / limit * 100,
                    ResetsAt = reset,
                });
            }
        }
        catch { }
        return windows;
    }

    /// Bool-or-number flag: true, 1 and "1" all count.
    private static bool GetFlag(JsonElement obj, string key)
    {
        if (obj.ValueKind != JsonValueKind.Object || !obj.TryGetProperty(key, out var v)) return false;
        return v.ValueKind switch
        {
            JsonValueKind.True => true,
            JsonValueKind.Number => v.GetDouble() != 0,
            JsonValueKind.String => v.GetString() == "1" || v.GetString() == "true",
            _ => false,
        };
    }

    // MARK: Credentials

    private static string ConfigPath => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".cursor", "cli-config.json");

    private static string? LoadToken()
    {
        if (_tokenCache != null) return _tokenCache;

        if (ReadConfigString("accessToken") is string fromConfig && fromConfig.Length > 0)
            return _tokenCache = fromConfig;

        foreach (var target in new[] { "cursor-access-token", "cursor-user" })
        {
            var value = CredentialManager.ReadGeneric(target)?.Trim();
            if (!string.IsNullOrEmpty(value))
                return _tokenCache = value;
        }
        return null;
    }

    /// The numeric Cursor userId lives in cli-config.json under authInfo.userId.
    private static string? LoadUserId()
    {
        if (_userIdCache != null) return _userIdCache;
        var id = ReadConfigString("userId");
        if (id != null) _userIdCache = id;
        return id;
    }

    private static string? ReadConfigString(string key)
    {
        try
        {
            if (!File.Exists(ConfigPath)) return null;
            using var doc = JsonDocument.Parse(File.ReadAllText(ConfigPath));
            if (!doc.RootElement.TryGetProperty("authInfo", out var auth)
                || auth.ValueKind != JsonValueKind.Object) return null;
            if (!auth.TryGetProperty(key, out var v)) return null;
            return v.ValueKind switch
            {
                JsonValueKind.String => v.GetString(),
                JsonValueKind.Number => ((long)v.GetDouble()).ToString(),
                _ => null,
            };
        }
        catch { return null; }
    }

    private static void ClearCache()
    {
        _tokenCache = null;
        _userIdCache = null;
    }
}

File: Sources/Brink/CursorProvider.swift
import Foundation

/// Reads the Cursor CLI's session JWT (macOS Keychain, service
/// "cursor-access-token" / account "cursor-user") and queries Cursor's
/// `usage-summary` endpoint — the source behind the "You've used N% of your
/// included usage" message in the Cursor app.
///
/// Design notes — same posture as ClaudeProvider:
/// - Auth is the browser-style session cookie `WorkosCursorSessionToken=
///   <userId>::<jwt>`. The numeric userId is read from ~/.cursor/cli-config.json
///   (written by `cursor-agent login`); the JWT comes from the Keychain.
/// - Brink only reads "cursor-access-token". The sibling "cursor-refresh-token"
///   is never touched — refreshing would rotate it and log the user out of
///   `cursor-agent`.
/// - Cursor exposes no public usage API; `usage-summary` is internal, so every
///   field is probed defensively and a missing shape degrades to a friendly
///   error rather than crashing.
/// - The server returns pre-computed percentages, so Brink displays them as-is.
struct CursorProvider: UsageProvider {
    let id = "cursor"

    private static let usageURL = URL(string: "https://cursor.com/api/usage-summary")!
    private static let keychainService = "cursor-access-token"
    private static let keychainAccount = "cursor-user"

    func fetch() async -> ProviderSnapshot {
        var snap = ProviderSnapshot(id: id, name: "Cursor",
                                    systemImage: "cursorarrow",
                                    windows: [], error: nil)
        guard let token = Self.loadToken(), let userId = Self.loadUserId() else {
            return ClaudeProvider.demoSnapshot(
                name: "Cursor", systemImage: "cursorarrow",
                note: L("Cursor CLI credentials not found — run `cursor-agent login`"))
        }

        do {
            var (data, status) = try await Self.requestUsage(token: token, userId: userId)
            if status == 401 || status == 403 {
                // cursor-agent rotated its token: drop cache, re-read once.
                Self.clearCache()
                if let fresh = Self.loadToken(), fresh != token {
                    (data, status) = try await Self.requestUsage(token: fresh, userId: userId)
                }
            }
            if status == 401 || status == 403 {
                snap.error = L("Unauthorized — run `cursor-agent login` to refresh")
                return snap
            }
            guard status == 200 else {
                snap.error = L("HTTP %d", status)
                return snap
            }
            snap.windows = Self.parseUsage(data)
            snap.updatedAt = Date()
            if snap.windows.isEmpty { snap.error = L("No usage data in response") }
            return snap
        } catch {
            snap.error = error.localizedDescription
            return snap
        }
    }

    private static func requestUsage(token: String, userId: String) async throws -> (Data, Int) {
        var request = URLRequest(url: usageURL)
        request.httpMethod = "GET"
        // Session-cookie auth: "WorkosCursorSessionToken=<userId>::<jwt>", ":" percent-encoded.
        request.setValue("WorkosCursorSessionToken=\(userId)%3A%3A\(token)",
                         forHTTPHeaderField: "Cookie")
        request.setValue("Brink/1.0", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = 15
        let (data, response) = try await URLSession.shared.data(for: request)
        return (data, (response as? HTTPURLResponse)?.statusCode ?? 0)
    }

    // MARK: Parsing

    /// usage-summary shape (fields probed defensively):
    /// { individualUsage: { plan: { totalPercentUsed, autoPercentUsed, apiPercentUsed,
    ///                              enabled, limit, used, remaining },
    ///                      onDemand: { enabled, limit, used } },
    ///   isUnlimited, billingCycleEnd }
    static func parseUsage(_ data: Data) -> [UsageWindow] {
        guard let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return []
        }
        let reset = ClaudeProvider.isoDate(root["billingCycleEnd"])

        // Unlimited plans have no meaningful percentage to show.
        if let unlimited = ClaudeProvider.number(root["isUnlimited"]), unlimited == 1 {
            return [UsageWindow(label: "Included usage", usedPercent: 0, resetsAt: reset)]
        }

        let individual = root["individualUsage"] as? [String: Any]
        let plan = individual?["plan"] as? [String: Any]
        var windows: [UsageWindow] = []

        // Headline window — total included usage (matches the app's banner).
        if let pct = ClaudeProvider.number(plan?["totalPercentUsed"]) {
            windows.append(UsageWindow(label: "Included usage", usedPercent: pct, resetsAt: reset))
        }
        if let pct = ClaudeProvider.number(plan?["autoPercentUsed"]) {
            windows.append(UsageWindow(label: "Auto", usedPercent: pct, resetsAt: reset))
        }
        if let pct = ClaudeProvider.number(plan?["apiPercentUsed"]) {
            windows.append(UsageWindow(label: "API", usedPercent: pct, resetsAt: reset))
        }

        // On-demand / usage-based spend, only when the user has enabled it.
        if let onDemand = individual?["onDemand"] as? [String: Any],
           ClaudeProvider.number(onDemand["enabled"]) == 1,
           let used = ClaudeProvider.number(onDemand["used"]),
           let limit = ClaudeProvider.number(onDemand["limit"]), limit > 0 {
            windows.append(UsageWindow(label: "On-demand",
                                       usedPercent: used / limit * 100, resetsAt: reset))
        }

        return windows
    }

    // MARK: Credentials

    private static var tokenCache: String?
    private static var userIdCache: String?
    private static var lastKeychainAttempt: Date?
    private static let keychainRetryInterval: TimeInterval = 10 * 60

    /// Lookup order, chosen so the macOS Keychain prompt appears at most once:
    ///  1) in-memory cache
    ///  2) Brink's own copy (mode 0600) — survives relaunches *and* rebuilds, which
    ///     matter here: an ad-hoc signature changes on every build and macOS then
    ///     treats the app as a different program, re-asking for Keychain access
    ///  3) the Keychain itself — the only step that can prompt
    /// The refresh token is never read; refreshing it would log the user out of
    /// `cursor-agent`.
    static func loadToken() -> String? {
        if let cached = tokenCache { return cached }

        if let data = try? Data(contentsOf: ownStoreURL),
           let token = String(data: data, encoding: .utf8)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
           !token.isEmpty {
            tokenCache = token
            return token
        }

        // Don't hammer the Keychain (and the user with prompts) if it keeps failing.
        if let last = lastKeychainAttempt, Date().timeIntervalSince(last) < keychainRetryInterval {
            return nil
        }
        lastKeychainAttempt = Date()

        guard let data = ClaudeProvider.keychainData(service: keychainService,
                                                     account: keychainAccount),
              let token = String(data: data, encoding: .utf8)?
                .trimmingCharacters(in: .whitespacesAndNewlines),
              !token.isEmpty else {
            return nil
        }
        tokenCache = token
        saveOwnCopy(token)
        return token
    }

    // MARK: Own token cache (~/Library/Application Support/Brink/cursor-token)

    static var ownStoreURL: URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            .appendingPathComponent("Brink", isDirectory: true)
            .appendingPathComponent("cursor-token")
    }

    static func saveOwnCopy(_ token: String) {
        let dir = ownStoreURL.deletingLastPathComponent()
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true,
                                                 attributes: [.posixPermissions: 0o700])
        try? Data(token.utf8).write(to: ownStoreURL, options: .atomic)
        try? FileManager.default.setAttributes([.posixPermissions: 0o600],
                                               ofItemAtPath: ownStoreURL.path)
    }

    /// The numeric Cursor userId lives in ~/.cursor/cli-config.json under
    /// authInfo.userId. Needed as the cookie's "<userId>::" prefix.
    static func loadUserId() -> String? {
        if let cached = userIdCache { return cached }
        let url = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".cursor/cli-config.json")
        guard let data = try? Data(contentsOf: url),
              let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let auth = root["authInfo"] as? [String: Any] else {
            return nil
        }
        let raw = auth["userId"]
        let id: String?
        if let s = raw as? String { id = s }
        else if let n = raw as? Int { id = String(n) }
        else if let d = raw as? Double { id = String(Int(d)) }
        else { id = nil }
        userIdCache = id
        return id
    }

    static func clearCache() {
        tokenCache = nil
        lastKeychainAttempt = nil
        try? FileManager.default.removeItem(at: ownStoreURL)
    }
}
